File: src/main/java/dev/salmon/keystrokes/hud/GuiKeyMouse.java
package dev.salmon.keystrokes.hud;

import dev.salmon.keystrokes.config.KeystrokesConfig;
import dev.salmon.keystrokes.hud.api.KeyRenderer;
import net.minecraft.client.settings.KeyBinding;

import java.util.LinkedList;
import java.util.Queue;

public class GuiKeyMouse extends GuiKey {
    private boolean pressed;
    private final Queue<Long> clicks = new LinkedList<>();

    public GuiKeyMouse(float x, float y, float width, float height, KeyBinding keyBinding) {
        super(x, y, width, height, keyBinding);
    }

    @Override
    public void render(float baseX, float baseY, float scale) {
        float x = baseX + relX * scale;
        float y = baseY + relY * scale;
        float w = width * scale;
        float h = height * scale;

        KeyRenderer.drawBackground(x, y, w, h, style, isPressed, percentFaded, scale);

        if (KeystrokesConfig.keystrokesElement.mouseCPS && getCps() > 0) {
                KeyRenderer.drawCenteredText(fr, this.getCps() + "", x, y, w, h, style, isPressed, percentFaded);
        } else {
            KeyRenderer.drawCenteredText(fr, getKeyName(), x, y, w, h, style, isPressed, percentFaded);
        }
    }

    @Override
    protected boolean isKeyDown(int code) {
        boolean down = super.isKeyDown(code) || pressed;
        if (pressed) pressed = false;
        return down;
    }

    public void pressed(int code) {
        if (code != this.keyBinding.getKeyCode()) return;
        clicks.add(System.currentTimeMillis() + 1000L);
        pressed = true;
    }

    public int getCps() {
        long time = System.currentTimeMillis();

        while (!clicks.isEmpty() && clicks.peek() < time) {
            clicks.remove();
        }

        return clicks.size();
    }
}

File: src/main/kotlin/org/polyfrost/keystrokes/hud/GuiKeyMouse.kt
package org.polyfrost.keystrokes.hud

import net.minecraft.client.settings.KeyBinding
import org.polyfrost.keystrokes.config.KeystrokesConfig.keystrokesElement
import org.polyfrost.keystrokes.hud.api.KeyRenderer.drawBackground
import org.polyfrost.keystrokes.hud.api.KeyRenderer.drawCenteredText
import java.util.*

class GuiKeyMouse(x: Float, y: Float, width: Float, height: Float, keyBinding: KeyBinding) : GuiKey(x, y, width, height, keyBinding) {
    private var pressed = false
    private val clicks: Queue<Long?> = LinkedList<Long?>()

    override fun render(baseX: Float, baseY: Float, scale: Float) {
        val x = baseX + relX * scale
        val y = baseY + relY * scale
        val w = width * scale
        val h = height * scale

        drawBackground(x, y, w, h, style, isPressed, percentFaded, scale)

        if (keystrokesElement.mouseCPS && this.cps > 0) {
            drawCenteredText(fr, this.cps.toString() + "", x, y, w, h, style, isPressed, percentFaded)
        } else {
            drawCenteredText(fr, keyName, x, y, w, h, style, isPressed, percentFaded)
        }
    }

    override fun isKeyDown(code: Int): Boolean {
        val down = super.isKeyDown(code) || pressed
        if (pressed) pressed = false
        return down
    }

    fun pressed(code: Int) {
        if (code != this.keyBinding.keyCode) return
        clicks.add(System.currentTimeMillis() + 1000L)
        pressed = true
    }

    val cps: Int
        get() {
            val time = System.currentTimeMillis()

            while (!clicks.isEmpty() && clicks.peek()!! < time) {
                clicks.remove()
            }

            return clicks.size
        }
}
